### ISSUE TITLE
Lentitud y retardos en la navegación de ALSIGM 3

### ISSUE THREAD

**Issue by fmoyanocr** [Cenatic/alsigm#26]
Buenas,

Desde la Diputación de Ciudad Real llevamos unos meses realizando la integración y migración de todas nuestras mejoras desde la versión 2.0 de ALSIGM hasta la versión 3.0.1 actual.

Siempre hemos apreciado que la aplicación iba más lenta pero hasta ahora no le habíamos dado importancia pues lo habíamos achacado a un tema de potencia de máquinas, memoria, optimización.

Sin embargo, nuestra sorpresa ha sido cuando hemos cambiado la máquina, hemos subido la memoria, hemos externalizado ciertos proyectos, o hemos desactivado mejoras como la auditoría, y no hemos apreciado mejora en la velocidad alguna.
Hasta ahora:
- Hemos probado con varios tomcats. Hemos visto claramente que tomcat 6 arranca mucho más rápido que el 7, pero no hay diferencias apreciables en la rapidez de la aplicación.
- Hemos probado con varias versiones de la BBDD Postgres, desde la 9.1 hasta las 8.3 que teníamos en SIGEM2.
- Hemos desactivado proyectos como el Scheduler o las Auditorías.
- Hemos revisado índices en la BBDD, comprobando que todos los nuevos índices que incorporamos a ALSIGM2 existen también en ALSIGM3.
- Hemos comprobado nuevas vistas del ALSIGM3 que no existían o cambiaban desde ALSIGM2, como SPAC_WL_PCD o SPAC_WL_CLOSE_TASK. Si bien estas vistas son un cuello de botella, sólo tienen importancia en la carga de la pantalla inicial y no en la navegación general.
- Hemos bajado el nivel de LOG.
- Hemos optimizado los recursos del tomcat, -XX:PermSize=4096m -XX:MaxPermSize=9216m
- Hemos optimizado la máquina, que ya tiene 16GB.
- Hemos optimizado recursos y configuración de la BBDD así como del Tomcat.
- Hemos llegado a conectar el Tomcat ALSIGM3 con una de las BBDD migrada de la máquina de ALSIGM2. Aunque no podemos descartar al 100% el tema de BBDD, creemos que no es el problema.
- Estamos empezando a mirar código, y hay ciertas clases como "LocaleFilter" que cantan desde el VisualVM.

Como todo el tema de las velocidades es muy subjetivo, hemos realizado un vídeo en el que se aprecia la diferencia clara de velocidad entre ambas versiones.
https://www.youtube.com/watch?v=ySR1RZGh9s0
El vídeo comienza con ALSIGM 2 y a partir del minuto 1.42 podéis apreciar la lentitud de nuestro ALSIGM 3.

Estamos ilusionados con la puesta en marcha en producción de ALSIGM 3 pero somos conscientes que no podemos propagar a los usuarios una aplicación que funciona entre 2 y 3 veces más lenta que la versión que se supone que mejora.
Rogamos que entre todos busquemos una solución y, los que ya tenéis integrado ALSIGM 3, nos comentéis si vuestra aplicación funciona más rápido que la nuestra.

Un saludo y muchas gracias,
Felipe

**Comment by moalmi1** [comment]
Hola Felipe,
En la UDC tenemos SIGEM 3.0.1, con Tomcat 7 y Oracle 11. La experiencia en cuanto a la velocidad de la aplicación es la misma que ofrecéis en el video. También estudiamos un poco el tema (mucho menos que vosotros, por lo que veo), aunque como es nuestra primera versión de SIGEM pensábamos que simplemente la aplicación era así. Viendo que versiones anteriores iban mucho más rápido, creo que sí que sería más que interesante encontrar una solución. 
Por si sirve de algo, en nuestras pruebas (ya digo que poco formales y hechas hace tiempo) nos pareció detectar que en las consultas a la base de datos en cada pantalla del registro se estaban obteniendo absolutamente todas las entidades del expediente (aunque no se estuvieran mostrando en ese momento) e incluso parecía que se repetían varias veces las mismas consultas para cada petición. 
Un saludo,Miguel A. Molinero
Date: Mon, 6 Jul 2015 05:18:32 -0700

**Comment by fmoyanocr** [comment]
Gracias por la información Miguel Ángel!
Muy interesante la información que me comentas del registro, lo investigaremos también.
Hoy vamos a conectarnos con San Fernando (Cádiz) para comprobar si es algo general de la versión 3.
Saludos,
Felipe

**Comment by moalmi1** [comment]
Hola Felipe,
Donde dije registro, quería decir tramitador. Disculpa por el lapsus :) 
Es decir, es en las mismas pantallas del tramitador que muestras en el video. Al acceder al expediente nos parece que se consultan todas sus entidades en cada una de las páginas (aunque solo se muestre una entidad).

Date: Tue, 7 Jul 2015 04:36:28 -0700

**Comment by fmoyanocr** [comment]
Sí, hemos visto algo, me imagino que te refieres al método loadScheme de la clase SchemeMgr, que es bastante costoso, y tiene un for anidado enorme.
Sin embargo, nos parece que la clase en ALSIGM2 es prácticamente igual que en ALSIGM3, aunque podemos volver a investigarlo.

También detectamos en su momento que con cada click se cargan todos los trámites y expedientes relacionados. Esto lo hemos solucionado tanto en ALSIGM2, como integrado en ALSIGM3.

Saludos

**Comment by moalmi1** [comment]
Hola Felipe,
Me refería a la segunda parte que comentas, la carga continua de entidades, trámites y, ahora que lo comentas, nos llamó la atención que también se cargaban siempre los expedientes relacionados.
En nuestro caso no hemos tocado nada de eso y la experiencia de uso en nuestro tramitador es similar  a la de vuestro video.
Si avanzáis algo en esta investigación sobre la lentitud del tramitador os agradeceríamos que no tengáis al tanto.
Gracias a vosotros!

Date: Tue, 7 Jul 2015 04:59:48 -0700

**Comment by carlosTisa** [comment]
Hola Felipe,
Debido a que las maquinas sobre las que nosotros ejecutamos SIGEM son bastante modestas y tienen ya sus años, no llegamos a experimentar un cambio significativo en cuanto al rendimiento de la aplicación desde que actualizamos a la versión 3.0.1. (hace más de un año)
Pero según vemos en el video, la versión 3.0.1 no nos va tan lenta como a vosotros, eso sí, tampoco nos iba tan rápido la versión 2 como a vosotros.
Si apreciamos que se ralentiza en exceso la aplicación cuando hacemos algún cambio de clase en caliente, pero se soluciona una vez reiniciada la máquina.
Por si os sirve de algo, nosotros ejecutamos directamente la aplicación desde la máquina virtual facilitada en la web, montada sobre un server2003 con 4Gb de RAM en un caso, y en un server2008 con 8Gb en otro.

Saludos!
